File: 269-alien-dictionary/269-alien-dictionary.java
class Dictionary {
    Map<Character, List<Character>> graph;
    boolean[] visited;
    boolean[] visiting;
    List<Character> alienWord;
    Dictionary() {
        graph = new HashMap<>();
        visited = new boolean[26];
        Arrays.fill(visited, false);
        visiting = new boolean[26];
        Arrays.fill(visiting, false);
        alienWord = new ArrayList<>(); 
    };
    
    private void getCharToGraph(String word) {
        for (char c: word.toCharArray()) {
            graph.computeIfAbsent(c, s -> new ArrayList<Character>());
        }
        
    }
    
    public boolean buildGraph(String[] words) {
        getCharToGraph(words[0]);
        for (int i = 0; i < (words.length-1); i++) {
            boolean isValid = buildDependency(words[i], words[i+1]);
            if (!isValid) return false;
        }
        return true;
        
    }
    
    private boolean buildDependency(String word1, String word2) {
        getCharToGraph(word1);
        getCharToGraph(word2);
        for (int i = 0; i < Math.min(word1.length(), word2.length()); i++) {
            char c1 = word1.charAt(i);
            char c2 = word2.charAt(i);
            if (c1 != c2) {
                graph.get(c2).add(c1);
                return true;
            }
        }
        if (word1.length() > word2.length()) {
            return false;
        } 
        return true;
        
    }
    
    public boolean buildAlienString(Character c) {
        
        visiting[c-'a'] = true;
        for (char nxtC: graph.get(c)) {
            if (visiting[nxtC - 'a']) return false;
            
            if (!visited[nxtC - 'a']) {
                boolean isCycle = buildAlienString(nxtC);
                if (!isCycle) return false;
            }   
        }
        // r -> e -> w -> alienWord = [w,e,r,t]
        visiting[c-'a'] = false;
        if (!visited[c-'a']) alienWord.add(c);
        visited[c-'a'] = true;
        
        return true;
    }
    

    //ftrew,     
}


class Solution {
    public String alienOrder(String[] words) {
        Dictionary dict = new Dictionary();
        boolean isValid = dict.buildGraph(words);
        if (!isValid) return "";
        //System.out.println(dict.graph);
        for (Map.Entry<Character, List<Character>> e: dict.graph.entrySet()) {
            isValid = dict.buildAlienString(e.getKey());
            if (!isValid) return "";
        }
        
        StringBuilder sb = new StringBuilder();
        for (Character c: dict.alienWord) {
            sb.append(c);
        }
        
        return sb.toString();
    }
}

File: 269-alien-dictionary/269-alien-dictionary.py
class AlienDictImpl:
    
    def __init__(self, words):
        self.graph = collections.defaultdict(list)
        self.words = words
        self.isValid = self.buildGraph()
        self.finalWord = []
        self.visited = {}
        
    def buildCharGraph(self, word):
        for c in word:
            if c not in self.graph: self.graph[c] = set()
        
    def buildOrder(self, word1, word2):
        self.buildCharGraph(word1)
        self.buildCharGraph(word2)
        
        for c1, c2 in zip(word1, word2):
            if c1 != c2:
                self.graph[c1].add(c2)
                return True
        if len(word1) > len(word2):
            return False
        return True
        
        
    def buildGraph(self):
        if len(self.words) == 1: self.buildCharGraph(self.words[0])
        for i in range(len(self.words)-1):
            isValid = self.buildOrder(self.words[i], self.words[i+1])
            if not isValid: return False
        
        return True
    
    def buildFinalWord(self, char):
        
        self.visited[char] = False
        # v = {a: False}
        # v = {a: F, c: F}
        
        # a c 
      
        isNxtChar = True
        for nxtChar in self.graph[char]:
            if nxtChar in self.visited:
                isNxtChar = self.visited[nxtChar]
                break
            isNotCycle = self.buildFinalWord(nxtChar)
            if not isNotCycle: return False
        self.finalWord.append(char)
        self.visited[char] = True
        # v = {a: True}
        return isNxtChar
            

class Solution:
    def alienOrder(self, words: List[str]) -> str:
        
        AlienDict = AlienDictImpl(words)
        if not AlienDict.isValid: return ""
        """
        self.graph = {
            "c": ["b"],
            "b": ["a"],
            "a": []
        }
        
        """
        for char, v in AlienDict.graph.items():
            if char not in AlienDict.visited:
                isValid = AlienDict.buildFinalWord(char)

            if not isValid: return ""
        
        AlienDict.finalWord.reverse()
        return "".join(AlienDict.finalWord)
            